Xilinx/XRT
# memory leakage in XRT library

Hi Xilinx,

I am using FFmpeg and XMA/XRT (2019.1) stack to encode video frames.

I used valgrind to check if the program has memory leakage problem, and finally I got below result:

==23700== 32 bytes in 1 blocks are definitely lost in loss record 2 of 20
==23700==    at 0x4C29F93: malloc (vg_replace_malloc.c:307)
==23700==    by 0x62CCA45: xma_enc_session_create (in /opt/xilinx/xrt/lib/libxmaapi.so.2.2.0)
==23700==    by 0x45A0F7: ??? (in /data/ffmpeg-3.2.2/ffmpeg/ffmpeg)
==23700==    by 0xA5546D: ??? (in /data/ffmpeg-3.2.2/ffmpeg/ffmpeg)
==23700==    by 0x499FD8: ??? (in /data/ffmpeg-3.2.2/ffmpeg/ffmpeg)
==23700==    by 0x47D938: ??? (in /data/ffmpeg-3.2.2/ffmpeg/ffmpeg)
==23700==    by 0x7096C04: (below main) (in /usr/lib64/libc-2.17.so)


==23700== 4,128 bytes in 1 blocks are definitely lost in loss record 18 of 20
==23700==    at 0x4C2C0A9: calloc (vg_replace_malloc.c:760)
==23700==    by 0x62D13E4: xma_res_alloc_kernel(void*, XmaSession*, XmaKernReq*, XmaKernType) (in /opt/xilinx/xrt/lib/libxmaapi.so.2.2.0)
==23700==    by 0x62CC912: xma_enc_session_create (in /opt/xilinx/xrt/lib/libxmaapi.so.2.2.0)
==23700==    by 0x45A0F7: ??? (in /data/ffmpeg-3.2.2/ffmpeg/ffmpeg)
==23700==    by 0xA5546D: ??? (in /data/ffmpeg-3.2.2/ffmpeg/ffmpeg)
==23700==    by 0x499FD8: ??? (in /data/ffmpeg-3.2.2/ffmpeg/ffmpeg)
==23700==    by 0x47D938: ??? (in /data/ffmpeg-3.2.2/ffmpeg/ffmpeg)
==23700==    by 0x7096C04: (below main) (in /usr/lib64/libc-2.17.so)

It seems that there is memory leakage within the XMA、 library.

How to reproduce:

1. Install memory leakage check tool: valgrind;
2. Run the tool when performing the encoding, for example:
valgrind --leak-check=full --show-leak-kinds=all --track-origins=yes --verbose --log-file=valgrind-out.txt ./ffmpeg -f rawvideo -s:v 416x240 -pix_fmt yuv420p -y -i ./RaceHorses_416x240_30.yuv -c:v xxx265 -b:v 1M -vf fps=30  race_horse_0.265
3. Check the memory leakage result.
